i6_models/config.py

```diff
@@ -51,3 +53,22 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class SubassemblyWithOptions:
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Callable[[ConfigType], ModuleType]
+    cfg: ConfigType
```

Comment:
I wonder if we really need to have this attrib, so that the cfg is nested here.

Originally, I thought that the config type itself would have the `module_class` attrib.

E.g. `ConformerConvolutionV1Config` below would just additionally have the `module_class = ConformerConvolutionV1`.

Or the `ConformerEncoderV1Config` would just have `module_class = ConformerEncoderV1`.

When setting config options in the encoder for the frontend, currently it would be like:
```python
conformer_cfg.frontend = SubassemblyWithOptions(module_class=MyFrontend, cfg=MyFrontendCfg(...))
```
But with what I suggest, it would just be:
```python
conformer_cfg.frontend = MyFrontendCfg(...)
```

Reply:
This is an option too ofc., I think @vieting proposed smth. similar. But i would not implement it in this here, because this requires a full new discussion. Here for now I would just go with how it was proposed in the meeting back when we laid out the structure of the conformer.

Reply:
Was this really proposed this way? Actually my understanding of what we discussed in the meeting was what I wrote here, not your variant.

Reply:
```
  conformer_cfg = ConformerOptions()
  conformer_cfg.frontend = SubassemblyWithOptions(repo.assemblies.VGGFrontend, vgg_opts)

```

This is in the meeting notes from 25.4.

Reply:
Ah ok, then I did not paid too much attention during that discussion. I would have voted for the easier variant otherwise. I would in general always vote for the easier solution. But ok then.

Reply:
We can change this of course. But I would not do this in this PR since after all this is about the conformer. 
And I would def. prefer a follow-up PR for this (or even one right now I can merge into this).

Reply:
Also, I think @curufinwe always wants to avoid any further discussion on things which have already been discussed once. Even if there were maybe misunderstandings. E.g. I still think we did some bad choices which make it all too complicated and difficult, not really usable, we could have made it much easier, more modular, more usable, and I'm not sure if everybody really understands this, esp the aspect on code complexity, but we don't want to discuss this anymore.

So basically, once there was some decision, this will not be changed anymore, at least for V1.

So it means, we would now stick with this.

Reply:
Agreed. In general I am not against rediscussing things that might have been bad decisions, here I don't want to discuss it though because this PR is somewhat unrelated.

> I don't expect that this will be changed now once we have it in master.

Why? If you think something is not good as it is you can open a PR/Issue and then we can rediscuss/change it.

Reply:
> > I don't expect that this will be changed now once we have it in master.
> 
> Why?

Just from experience. Also that was my understanding on the decisions I'm unhappy with (complicated, difficult, complex, non-modular code design). I'm happy to discuss discuss any of that again.

Maybe the situation in my issue is a bit different, because those things are versionized (`ConformerConvolutionV1`), and as I understand, we would want to just have `ConformerConvolutionV2` for any future easier cleaner code. I just wondered why not directly have the easier and cleaner code from the beginning.

And here in this case, this is not versionized. So any change on this would affect other code which uses it.
Or maybe it should be versionized as well? `SubassemblyWithOptionsV1`? But maybe not.

Maybe we can stop the discussion then here in this PR on this.